scipy/signal/ltisys.py

```diff
@@ -2494,3 +2496,229 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
     full_state_feedback.nb_iter = nb_iter
 
     return full_state_feedback
+
+
+def controller_hessenberg(A, B, C=None):
+    r"""
+    Reduce the linear system :math:`AX+BU` to its controller Hessenberg form.
+    
+    Find an orthogonnal similarity transformation P and reduce the linear
+    system :math:`AX+BU` to the controller Hessenberg form 
+    :math:`\bar A, \bar B, \bar C` (Abar, Bbar, Cbar) where:
+    
+    .. math::
+        \begin{array}{lcl} P A P^T &=& \bar A \\
+                            P B &=& \bar B\end{array}
+
+    The new system has the following form:
+    
+    .. math:: 
+
+        \bar A = \left[ \begin{array}{cc}
+                        H_c & H_{xx} \\
+                        0 & H_{nc}
+                        \end{array} \right]
+
+    :math:`\bar A` is a block upper Hessenberg matrix hence the `H` prefix used
+    for its blocks.
+    
+    .. math:: 
+
+        \bar B = \left [ \begin{array}{c} B_c \\ 0 \end{array} \right ]
+
+        
+    :math:`H_c, B_c` is the controllable part of the original system and
+    :math:`AX+BU` is controllable if :math:`\bar A` is of full rank. If the
+    real part of the eigenvalues of :math:`H_{nc}` are negative the system
+    will be stabilizable.
+        
+    Parameters
+    ----------
+    A, B : ndarray
+        State-space representation of linear system :math:`AX+BU`.
+        
+    C : ndarray, optionnal
+        The output matrix of the linear system (i.e :math:`Y=CX`).
+        Default is None.
+                
+    Returns
+    -------
+    Abar, Bbar : ndarray
+        The controller Hessenberg form of the linear system :math:`AX+BU`
+
+    Cbar : ndarray    
+        Optionnaly if C (output matrix) is given the function will also return
+        :math:`\bar C = C P` otherwise `Cbar = None` will be returned.
+        Cbar has the following form:
+      
+         .. math::
+             \bar C = \left[ \begin{array}{cc}
+                             C_c & C_{nc}
+                             \end{array} \right]
+
+        With :math:`C_c` such that the transfer function of (A,B,C) and
+        (Hc,Bc,Cc) are the same.
+    
+    P : ndarray
+        The orthogonnal similarity transformation
+        
+    control_index : int
+        The controllability index, i.e the number of controllable states, of
+        the linear system :math:`AX+BU`. If ``control_index==A.shape[0]`` then
+        the linear system is controllable. Control_index is also the rank of 
+        :math:`\bar A`.
+        
+    Notes
+    -----
+    The algorithm comes from [1]_. The output from this function has been tested
+    with random controllable and uncontrollable system against Octave which
+    uses tb01ud from Slicot and all results were similar using the default
+    tolerances of `allcose`.
```

Comment:
I'd move this note about "tested against ...." to the unit tests. We regularly test against other software (R, Matlab, Octave, etc.) and normally don't mention this in a docstring.